linux: fix cross

alyssais opened the pull request:
The .unwrapped attributes are not correctly spliced.  Use the top level attributes instead, which are spliced correctly.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

alyssais committed:
rustc-unwrapped: add attribute

rustc.unwrapped is available, but it isn't spliced correctly.  Having
this attribute available at the top level is consistent with
e.g. rust-bindgen-unwrapped.

alyssais committed:
linux: fix cross

The .unwrapped attributes are not correctly spliced.  Use the top
level attributes instead, which are spliced correctly.

Fixes: e9b2edb49163 ("linux: un-simplify toolchain selection")

alyssais committed:
linux: use consistent rustc and rust-bindgen

If we're going to use the unwrapped versions, we should do so
consistently in both parts of the kernel build, not mix and match.

Fixes: e9b2edb49163 ("linux: un-simplify toolchain selection")

georgewhewell commented:
i think something is broken-

```
❯ nix build github:alyssais/nixpkgs/linux-cross#legacyPackages.x86_64-linux.pkgsCross.gnu64.linuxPackages_latest.kernel
error: builder for '/nix/store/nxyhlfp1r7rm0m16ykwrbd1b8sdpp0pj-linux-config-6.16.drv' failed with exit code 255;
       last 25 log lines:
       > GOT: * Runtime Testing
       > GOT: *
       > QUESTION: Runtime Testing, NAME: RUNTIME_TESTING_MENU, ALTS: N/y/?, ANSWER: n
       > GOT: n
       > QUESTION: Memtest, NAME: MEMTEST, ALTS: Y/n/?, ANSWER: y
       > GOT: y
       > QUESTION: Microsoft Hyper-V driver testing, NAME: HYPERV_TESTING, ALTS: N/y/?, ANSWER:
       > GOT:
       > GOT: *
       > GOT: * Rust hacking
       > GOT: *
       > GOT: #
       > GOT: # configuration written to .config
       > GOT: #
       > GOT: make[1]: Leaving directory '/build/linux-6.16/build'
       > GOT: make: Leaving directory '/build/linux-6.16'
       > warning: unused option: CRC32_SELFTEST
       > warning: unused option: CRYPTO_TEST
       > error: unused option: DRM_PANIC_SCREEN_QR_CODE
       > warning: unused option: POWER_RESET_GPIO
       > warning: unused option: POWER_RESET_GPIO_RESTART
       > warning: unused option: REISERFS_FS_POSIX_ACL
       > warning: unused option: REISERFS_FS_SECURITY
       > warning: unused option: REISERFS_FS_XATTR
       > error: unused option: RUST
       For full logs, run:
         nix log /nix/store/nxyhlfp1r7rm0m16ykwrbd1b8sdpp0pj-linux-config-6.16.drv
error: 1 dependencies of derivation '/nix/store/j3n421r34v43g4bjj0psjl4hm155xvlb-linux-6.16.drv' failed to build
```
it used to work
```
❯ nix build github:nixos/nixpkgs/nixos-unstable-small#legacyPackages.x86_64-linux.pkgsCross.gnu64.linuxPackages_latest.kernel --print-out-paths
/nix/store/7bh7fkf277bm75s34r12v8jarsandhzr-linux-6.16
```

alyssais committed:
rust-bindgen: link directly with libclang

There's not really any reason for bindgen to dlopen libclang at all in
Nixpkgs.  Moving this out of the wrapper means that it will still work
even when we don't want all the standard Nixpkgs cflags behavior, like
in the Linux kernel build.

ccicnce113424 commented:
I think unwrapped bindgen is missing some libs.
```console
$ make rustavailable
***
*** Running 'bindgen' to check the libclang version (used by the Rust
*** bindings generator) failed with code 101. This may be caused by
*** a failure to locate libclang. See output and docs below for details:
***

thread 'main' panicked at /build/rust-bindgen-unwrapped-0.72.0-vendor/bindgen-0.72.0/lib.rs:616:27:
Unable to find libclang: "couldn't find any valid shared libraries matching: ['libclang.so', 'libclang-*.so', 'libclang.so.*', 'libclang-*.so.*'], set the `LIBCLANG_PATH` environment variable to a path where one of these files can be found (invalid: [])"
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
***
***
*** Please see Documentation/rust/quick-start.rst for details
*** on how to set up the Rust support.
***
```

alyssais commented:
Yes, it is. I'm looking into it. Adding `rustPlatform.bindgenHook` fixes it, but ideally bindgen would just link the library…
